container/client.go

```diff
@@ -0,0 +1,266 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package container
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/fsouza/go-dockerclient"
+	"github.com/intelsdi-x/snap-plugin-collector-docker/config"
+)
+
+const (
+	dockerVersionKey string = "Version"
+)
+
+// DockerClientInterface provides methods i.a. for interaction with the docker API.
+type DockerClientInterface interface {
+	ListContainersAsMap() (map[string]*ContainerData, error)
+	InspectContainer(string) (*docker.Container, error)
+	FindCgroupMountpoint(string, string) (string, error)
+	FindControllerMountpoint(string, string, string) (string, error)
+	GetDockerParam(string) (string, error)
+}
+
+// DockerClient holds fsouza go-dockerclient instance ready for communication with the server endpoint `unix:///var/run/docker.sock`,
+// cache instance which is used to store output from docker container inspect (to avoid execute inspect request multiply times, it is called only once per container)
+// and diskUsageCollector which is responsible for collecting container disk usage (based on `du -u` command) in the background
+type DockerClient struct {
+	cl           *docker.Client
+	inspectCache map[string]*docker.Container
+	inspectMutex sync.Mutex
+}
+
+type deviceInfo struct {
+	device string
+	major  string
+	minor  string
+}
+
+// NewDockerClient returns dockerClient instance ready for communication with the server endpoint `unix:///var/run/docker.sock`
+func NewDockerClient(endpoint string) (*DockerClient, error) {
+	client, err := docker.NewClient(endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot initialize docker client instance with the given endpoint `%s`, err=%v", endpoint, err)
+	}
+
+	err = client.Ping()
+	if err != nil {
+		return nil, err
+	}
+
+	dc := &DockerClient{
+		cl:           client,
+		inspectCache: map[string]*docker.Container{},
+	}
+
+	config.DockerVersion, err = dc.version()
+	if err != nil {
+		return nil, err
+	}
+
+	return dc, nil
+}
+
+// InspectContainer returns details information about running container
+func (dc *DockerClient) InspectContainer(id string) (*docker.Container, error) {
+	dc.inspectMutex.Lock()
+	defer dc.inspectMutex.Unlock()
+
+	// check if the inspect info is already stored in inspectCache
+	if info, haveInfo := dc.inspectCache[id]; haveInfo {
+		return info, nil
+	}
+
+	info, err := dc.cl.InspectContainer(id)
+	if err != nil {
+		return nil, err
+	}
+	dc.inspectCache[id] = info
+
+	return info, nil
+}
+
+// GetDockerParam returns given parameter value fro running docker engine
+func (dc *DockerClient) GetDockerParam(param string) (string, error) {
+	env, err := dc.cl.Info()
```

Comment:
What do you think about changing it to GetDockerParam**s** and allowing to make a request for a few params, so `dc.cl.Info()` will be executed only once. Does it have a sense?